Kubevirt - Node Selectors Modal

glekner opened the pull request:
based upon #4603

glekner commented:
- #4603 is merged
- Removed WIP label
- Node search is debounced to every .5 sec
- There is a custom design by me right now ( gif below ) , probably will change in the future.

@suomiy @yaacov @irosenzw  can you review?

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/24938324/75987417-91d86600-5ef8-11ea-8d04-7a02fec098d3.gif)

glekner commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/node-selector-modal/index.tsx:
```diff
@@ -0,0 +1,109 @@
+  const { vmLikeEntity, nodeSelector = {}, handlePromise, setOpen, ...restProps } = props;
+
+  const [selector, setSelector] = React.useState({
+    ...Object.entries(nodeSelector).map(([key, value]) => ({ key, value })),
```
I want to numerate the labels and control their fields, right now they don't have fixed fields `key ,value`
they have e.g. `{ performance: 'high', cpumanager: 'true' }`
My code:

```js
{ performance: 'high', cpumanager: 'true' }  => 
{ 0: { key: 'performace', value: 'high' }, 1: { key: 'cpumanager', value: 'true' }`
```
this way i can edit they `key, value` and fetch a specific one by its index
your suggestions just leaves the object as it is.

glekner commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/node-selector-modal/index.tsx:
```diff
@@ -0,0 +1,109 @@
+
+  const [isLoading, setIsLoading] = React.useState<boolean>(true);
+  const [shouldPatch, setShouldPatch] = React.useState<boolean>(false);
+  const [showPatchError, setPatchError] = React.useState<boolean>(false);
```
I use the PF4 Modal, not the factory `react-modal`. we began using it in the boot order Modal, so we need an extra boolean for the error

suomiy replied:
Can we use the createModalLauncher approach as there are benefits of having new instance for each Dialog run? 
Plus it is a standard way of creating modals in the console.

glekner commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/node-selector-modal/index.tsx:
```diff
@@ -0,0 +1,109 @@
+
+  const onClose = () => {
+    setIsLoading(true);
+    setSelector({
```
Same, the state remains with the new PF4 modal, so we need to reset it.

glekner commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/shared/LabelsList/labels-list.tsx:
```diff
@@ -0,0 +1,72 @@
+  onLabelChange,
+  onLabelDelete,
+  addRowText = SCHEDULING_ADD_LABEL,
+  showEffect = false,
```
In the upcoming Tolerations Modal, will create it after this PR

suomiy replied:
Can you please remove it and move it to the Tolerations PR? 
It is easier for review and one can see the relevant code which will be using it in one place/

glekner commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/node-selector-modal/node-selector-modal.tsx:
```diff
@@ -0,0 +1,103 @@
+
+  const [qualifiedNodes, setQualifiedNodes] = React.useState([]);
+  // Node search 1 sec after user stopped typing
+  React.useEffect(() => {
```
I wanted to slowdown the array searching yes. I thought about cases with 1000's of nodes.
maybe theres a better way to throttle it? it works pretty good as it is.
I don't think we should fire a function after every character the user types. we should wait for at least 1 sec after user stopped typing

suomiy replied:
why should we wait after the user stopped typing? it is just a simple filter of data we already have. I don't see any benefit for the user.

glekner replied:
removed timeout

glekner commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/node-selector-modal/index.tsx:
```diff
@@ -0,0 +1,109 @@
+
+  const onSubmit = async () => {
+    if (shouldPatch) {
+      const k8sSelector = Object.assign(
```
this turns https://github.com/openshift/console/pull/4614/files#r389737953
back to the original prop

suomiy replied:
ok, but still I would rather see object spread instead of assign - easier to read

suomiy replied:
sorry, I misread the code last time. In that case can we use reduce for better readability and less object creations?

glekner commented on frontend/packages/kubevirt-plugin/src/k8s/wrapper/vm/vmi-wrapper.ts:
```diff
@@ -50,4 +50,6 @@ export class VMIWrapper extends K8sResourceWrapper<VMIKind> implements VMILikeMe
 
   isDedicatedCPUPlacement = () => this.data.spec?.domain?.cpu?.dedicatedCpuPlacement || false;
+
+  getNodeSelector = () => this.data.spec?.template?.spec?.nodeSelector;
```
`getNodeSelector` doesn't accept `VMI`, whats the best solution here?

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/node-selector-modal/node-selector-modal.tsx:
```diff
@@ -0,0 +1,104 @@
+          value: 'true',
+        });
+
+        const newNodes = loadedNodes.filter((node) =>
```
you can getLabels and check if they are not null first and then reuse the labels object for getting the value by key

glekner commented:
- removed timeout delay
- removed Toleration modal's infrastructure for later PR
- implemented @suomiy suggestions.

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/shared/LabelRow/label-row.scss:
```diff
@@ -0,0 +1,12 @@
+.kv-label__row {
```
do we have to use css here? can't we rather use https://patternfly-react.surge.sh/documentation/react/layouts/grid or https://patternfly-react.surge.sh/documentation/react/layouts/split or some other layouts

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/shared/LabelsList/labels-list.scss:
```diff
@@ -0,0 +1,16 @@
+.kv-labels-list__column-title {
```
same: do we have to use css here? can't we rather use https://patternfly-react.surge.sh/documentation/react/layouts/grid or https://patternfly-react.surge.sh/documentation/react/layouts/split or some other layouts

matthewcarleton replied:
ya I agree, this doesn't feel like something that needs custom css. I think we've already build very similar layouts in the console already like this right?

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/shared/NodeChecker/node-checker.scss:
```diff
@@ -0,0 +1,21 @@
+.kv-node-checker__container {
```
same here

glekner commented on frontend/packages/kubevirt-plugin/src/k8s/patches/vm/vm-scheduling-patches.ts:
```diff
@@ -0,0 +1,16 @@
+): Patch[] =>
+  getVMLikePatches(vmLikeEntity, () => [
+    new PatchBuilder('/spec/template/spec/nodeSelector')
+      .setOperation(PatchOperation.REPLACE)
```
fixed

glekner commented on frontend/packages/kubevirt-plugin/src/selectors/vm/selectors.ts:
```diff
@@ -149,3 +149,8 @@ export const getStorageClassNameByDisk = (vm: VMKind, diskName: string) =>
     getDataVolumeTemplates(vm).find((vol) => getName(vol).includes(diskName)),
   );
+
+export const getNodeSelector = (vm: VMILikeEntityKind) =>
```
the paths seem ok right? VMI's spec is right after `vm.`

suomiy replied:
yes

glekner commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/node-selector-modal/node-selector-modal.tsx:
```diff
@@ -0,0 +1,79 @@
+        key: SCHEDULING_REQUIRED_LABEL,
+        value: 'true',
+      });
+      const newNodes = loadedNodes.filter((node) =>
```
fixed

glekner commented:
- fixed 2 comments @suomiy 
waiting for css review now

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/node-selector-modal/node-selector-modal.tsx:
```diff
@@ -0,0 +1,84 @@
+      const newNodes = [];
+      loadedNodes.forEach((node) => {
+        const nodeLabels = getLabels(node);
+        if (labels.every(({ key, value }) => nodeLabels[key] === value)) {
```
```suggestion
        if (nodeLabels && labels.every(({ key, value }) => nodeLabels[key] === value)) {
```

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/node-selector-modal/index.tsx:
```diff
@@ -8,20 +8,30 @@ import {
 const NodeSelectorModal: React.FC<NodeSelectorModalProps> = (props) => {
   const { vmLikeEntity, nodeSelector = {}, handlePromise, cancel, close, ...restProps } = props;
 
+  const [initialNodeSelector, setInitialNodeSelector] = React.useState<NodeSelector>(nodeSelector);
```
this nodeSelector should come from the VM, not from the props

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/node-selector-modal/index.tsx:
```diff
@@ -8,20 +8,30 @@ import {
     Object.entries(nodeSelector).map(([key, value], id) => ({ id, key, value })),
   );
 
+  const onReload = (vm: VMLikeEntityKind) => {
```
can we pass just the nodeSelector and not entire entity?

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/node-selector-modal/index.tsx:
```diff
@@ -8,20 +8,30 @@ import {
+    const updatedSelector = getNodeSelector(vm);
+    setInitialNodeSelector(updatedSelector);
+    setSelectorLabels(
+      Object.entries(updatedSelector || {}).map(([key, value], id) => ({ id, key, value })),
```
can we make and reuse a function for this?

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/node-selector-modal/node-selector-modal.tsx:
```diff
@@ -1,28 +1,37 @@
 
 export const NSModal = ({
   nodes,
+  nodeSelector,
```
can we name it `initialNodeSelector` also here?

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/node-selector-modal/index.tsx:
```diff
@@ -0,0 +1,113 @@
+      handlePromise(
+        k8sPatch(
+          getVMLikeModel(vmLikeEntity),
+          vmLikeEntity,
```
vmLikeFinal should be used in the submit request as well - or we might reach inconsistencies

suomiy replied:
It would be probably much easier to keep the initialNodeSelector state and the callback inside NSModal and use this one only as a resource wrapper - so we will have the recent VM in the scope

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/node-selector-modal/node-selector-modal.tsx:
```diff
@@ -0,0 +1,126 @@
+        onSubmit={onSubmit}
+        onCancel={onClose}
+        submitButtonText="Apply"
+        infoTitle={showCollisionAlert && 'Node Selector has been updated outside this flow.'}
```
I am also getting this alert for missing node selector

suomiy commented on frontend/packages/kubevirt-plugin/src/components/vm-templates/vm-template-resource.tsx:
```diff
@@ -149,9 +150,20 @@ export const VMTemplateSchedulingList: React.FC<VMTemplateResourceSummaryProps>
+        title={NODE_SELECTOR_MODAL_TITLE}
+        idValue={prefixedID(id, 'node-selector')}
+        editButtonId={prefixedID(id, 'node-selectors-edit')}
+        onEditClick={() => nodeSelectorModal({ vmLikeEntity: template, nodeSelector })}
```
missing the blocking prop (: , plus in the vm

matthewcarleton commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/shared/NodeChecker/node-checker.tsx:
```diff
@@ -0,0 +1,57 @@
+        </div>
+      </Popover>
+      <Text component={TextVariants.h4}>
+        {isLoading ? ' Searching for Nodes' : ' qualify for this node selector'}
```
Before the users clicks add label we shouldn't see this message. There is no selector yet. I would suggest saying "Add a Node selector label to see qualifying nodes" 
Also are you going to use PF inline notifications here? I Think the blue/grey is problematic especially without a timeout. You don't really notice it changing.

glekner commented:
- Created `useLabels`, `useNodeQualifier`, `useCollisionChecker` Hooks for reusability, please review @suomiy 
- Changed empty state add label text to `Add a Node selector label to see qualifying nodes`
- Changed Node checker to an inline alert
- Removed all custom css but the external link, they don't have their own for some reason and they overlap rows. see below

<img width="291" alt="Screen Shot 2020-03-12 at 3 15 20" src="https://user-images.githubusercontent.com/24938324/76478655-0e47d900-6412-11ea-89e4-3494387a41ff.png">

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/shared/hoc.ts:
```diff
@@ -0,0 +1,97 @@
+  vmLikeEntity: VMLikeEntityKind,
+  vmLikeEntityLoading: FirehoseResult<VMLikeEntityKind>,
+  getEntity: (vm: VMLikeEntityKind) => T,
+  setEntity: React.Dispatch<React.SetStateAction<T>>,
```
it does too many things. setEntity should be a responsibility of the caller of reload

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/shared/hoc.ts:
```diff
@@ -0,0 +1,97 @@
+      setCollisionAlert(true);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [vmLikeEntityLoading]);
```
+ here for deps

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/shared/hoc.ts:
```diff
@@ -0,0 +1,97 @@
+};
+
+export const useCollisionChecker = <T>(
+  vmLikeEntity: VMLikeEntityKind,
```
I think you can resolve the vmLikeFinal before calling useCollisionChecker and pass it here

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/shared/hoc.ts:
```diff
@@ -0,0 +1,97 @@
+export const useCollisionChecker = <T>(
+  vmLikeEntity: VMLikeEntityKind,
+  vmLikeEntityLoading: FirehoseResult<VMLikeEntityKind>,
+  getEntity: (vm: VMLikeEntityKind) => T,
```
we could make this function generic and move it to hooks - we could replace this with isEqual function. Meaning

this method would have only two parameters 

- `entity: T` // which is our VM
- `isEqual: (initialEntity: T, newEntity: T) => boolean `

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/shared/hoc.ts:
```diff
@@ -0,0 +1,97 @@
+  const [entity, setInitialEntity] = React.useState<T>(getEntity(vmLikeEntity));
+  const [showCollisionAlert, setCollisionAlert] = React.useState<boolean>(false);
+
+  const onReload = () => {
```
also + useCallback

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/shared/hoc.ts:
```diff
@@ -0,0 +1,97 @@
+
+export const useNodeQualifier = <T extends IDLabel>(
+  initialLabels: T[],
+  nodes: FirehoseResult<VMLikeEntityKind[]>,
```
wrong type

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/shared/hoc.ts:
```diff
@@ -0,0 +1,97 @@
+import { IDLabel } from './types';
+
+export const useNodeQualifier = <T extends IDLabel>(
+  initialLabels: T[],
```
they are not initial - they will be changed in the future

glekner commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/shared/LabelsList/labels-list.scss:
```diff
@@ -0,0 +1,8 @@
+.kv-scheduling-labels__grid {
+  grid-gap: var(--pf-global--spacer--md);
```
the gutter "sm" prop doesn't work for me..tried all of them and they produce the same result

glekner commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/shared/LabelsList/labels-list.tsx:
```diff
@@ -0,0 +1,66 @@
+  addRowText = SCHEDULING_ADD_LABEL,
+}) => (
+  <Grid className="kv-scheduling-labels__grid">
+    {labels.length > 0 && [
```
it looks weird having the headers with no inputs..

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/shared/hooks.ts:
```diff
@@ -0,0 +1,33 @@
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [labels, loadedNodes]);
+
+  return [qualifiedNodes, loadError];
```
loadError is just a pasthrough value. The callee can call getLoadError itself.

suomiy commented on frontend/packages/kubevirt-plugin/src/components/LabelsList/labels-list.tsx:
```diff
@@ -0,0 +1,74 @@
+
+export const LabelsList = <T extends IDLabel>({
+  labels,
+  onLabelAdd,
```
it is much safer - also type-wise for the caller to pass a function like this

`onLabelAdd = () => add({ key: '', value: '' })`

plus no need for the id - as it will be supllied

suomiy replied:
can we make it generic and not pass newLabel?

glekner commented on frontend/packages/kubevirt-plugin/src/components/LabelsList/labels-list.tsx:
```diff
@@ -0,0 +1,74 @@
+    <Grid className="kv-scheduling-labels__grid">
+      {labels.length > 0 && [
+        <React.Fragment key="label-title-row">
+          <GridItem span={7}>
```
keys are usually longer than values, so the input should be wider

suomiy commented on frontend/packages/kubevirt-plugin/src/hooks/use-collision-checker.ts:
```diff
@@ -0,0 +1,22 @@
+    if (!isEqual(initialEntity, entity)) {
+      setCollisionAlert(true);
+    }
+  }, [initialEntity, entity, isEqual]);
```
probably not isEqual as the user will forget to call React.useCallback on it

suomiy replied:
also it probably doesnt make sense for isEqual to change over time

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/shared/types.ts:
```diff
@@ -0,0 +1,3 @@
+import { IDLabel } from '../../../../hooks/use-id-labels';
+
+export type NodeSelectorLabel = IDLabel;
```
does it makes sense to have this type afterall?

suomiy commented on frontend/packages/kubevirt-plugin/src/components/LabelsList/labels-list.tsx:
```diff
@@ -0,0 +1,85 @@
+import { IDLabel } from './types';
+import './labels-list.scss';
+
+export const LabelsList = <T extends IDLabel>({
```
you can default the type to IDLabel here

glekner replied:
how would you do that?

suomiy commented on frontend/packages/kubevirt-plugin/src/components/LabelsList/LabelRow/label-row.tsx:
```diff
@@ -0,0 +1,45 @@
+import { GridItem, TextInput, Button } from '@patternfly/react-core';
+import { IDLabel } from '../types';
+
+export const LabelRow = <T extends IDLabel>({ label, onChange, onDelete }: LabelRowProps<T>) => {
```
+here

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/shared/hooks.ts:
```diff
@@ -0,0 +1,32 @@
+import { getLoadedData, isLoaded } from '../../../../utils';
+import { IDLabel } from '../../../LabelsList/types';
+
+export const useNodeQualifier = <T extends IDLabel>(
```
\+ here

suomiy commented on frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/node-selector-modal/node-selector-modal.tsx:
```diff
@@ -0,0 +1,128 @@
+
+    const onSubmit = async () => {
+      const k8sSelector = selectorLabels.reduce(
+        (acc, { key, value }) => ({ ...acc, [key]: value }),
```
nit: you can reuse the acc here

suomiy commented:
/lgtm
